src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -109,20 +109,127 @@ static CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE spvcrossType2E_TYPE
+            &m_pushConstantBytes[pc.offset],
+            &m_pushConstantBytes[pc.offset + pc.size]
+        );
+
+        // iterate over all bytes used
+        const IShader::E_SHADER_STAGE shaderStage = stageData->getParams().shader->getStage();
+        for (auto it = pcRangesSpan.begin(); it != pcRangesSpan.end(); ++it)
+            *it |= shaderStage;
+    }
+
+    return true;
 }
 
 //
 NBL_API2 core::smart_refctd_dynamic_array<SPushConstantRange> CSPIRVIntrospector::CPipelineIntrospectionData::createPushConstantRangesFromIntrospection(core::smart_refctd_ptr<const CStageIntrospectionData>& introspection)
 {
-    auto& ps = introspection->getPushConstants();
-    auto a = ps.type->memberInfoStorage;
+    auto& pc = introspection->getPushConstants();
 
-    return nullptr;
+    core::vector<SPushConstantRange> tmp; 
+    tmp.reserve(MaxPushConstantsSize);
+    
+    SPushConstantRange prev = {
+        .stageFlags = m_pushConstantBytes[0].value,
+        .offset = 0,
+        .size = 0
+    };
+
+    SPushConstantRange current = {
+        .offset = 0,
+        .size = 0
+    };
+
+    // TODO: test
+    // run-length encode m_pushConstantBytes
+    for (uint32_t currentByteOffset = 1u; currentByteOffset < MaxPushConstantsSize; ++currentByteOffset)
+    {
+        current.stageFlags = m_pushConstantBytes[currentByteOffset].value;
+        current.offset++;
+        current.size++;
+
+        if (current.stageFlags != prev.stageFlags)
+        {
+            if (prev.stageFlags)
+            {
+                prev.size = current.offset - prev.offset;
+                tmp.push_back(prev);
+            }
+            prev = current;
+        }
+    }
+
+    if (prev.stageFlags)
+    {
+        prev.size = current.offset - prev.offset;
+        tmp.push_back(prev);
+    }
+
+    if(tmp.size() == 0)
+        return nullptr;
+
+    std::ostringstream debug;
+    for (auto it = tmp.begin(); it != tmp.end(); it++)
+        debug << "Stage flags: " << std::bitset<32>(it->stageFlags) << " offset: " << it->offset << " size: " << it->size << std::endl;
+    std::cout << debug.str();
+
+    return core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<SPushConstantRange>>(tmp);
 }
 NBL_API2 core::smart_refctd_ptr<ICPUDescriptorSetLayout> CSPIRVIntrospector::CPipelineIntrospectionData::createApproximateDescriptorSetLayoutFromIntrospection(const uint32_t setID)
 {
-    return nullptr;
+    std::vector<ICPUDescriptorSetLayout::SBinding> outBindings;
+    outBindings.reserve(m_descriptorSetBindings[setID].size());
+    std::ostringstream debug;
+
+    for (const auto& binding : m_descriptorSetBindings[setID])
+    {
+        auto& outBinding = outBindings.emplace_back();
+
+        outBinding.binding = binding.binding;
+        outBinding.count = binding.count;
+        outBinding.type = binding.type;
+        outBinding.stageFlags = binding.stageMask;
+        // TODO outBinding.createFlags = 
+
```

Comment:
sane default would be `ECF_NONE` we can't infer anything about them from the SPIR-V /  shader